pkg/operator/csidriveroperator/driverstarter.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/openshift/library-go/pkg/controller/manager"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/staticresourcecontroller"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/client-go/dynamic"
+)
+
+const (
+	conditionsPrefix = "CSIDriverStarter"
+	infraConfigName  = "cluster"
+)
+
+// This CSIDriverStarterController starts CSI driver controllers based on the underlying cloud.
+// It does not install anything by itself, only monitors Infrastructure
+// instance and starts ControllerManagers for the particular cloud.
+// It produces following Conditions:
+// CSIDriverStarterDegraded - error checking the Infrastructure
+type CSIDriverStarterController struct {
+	operatorClient v1helpers.OperatorClient
+	infraLister    openshiftv1.InfrastructureLister
+	versionGetter  status.VersionGetter
+	targetVersion  string
+	eventRecorder  events.Recorder
+
+	controllers []csiDriverControllerManager
+}
+
+type csiDriverControllerManager struct {
+	name           string
+	operatorConfig csioperatorclient.CSIOperatorConfig
+	platform       configv1.PlatformType
+	mgr            manager.ControllerManager
+	running        bool
+}
+
+func NewCSIDriverStarterController(
+	clients *csoclients.Clients,
+	dynamicClient dynamic.Interface,
+	resyncInterval time.Duration,
+	versionGetter status.VersionGetter,
+	targetVersion string,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &CSIDriverStarterController{
+		operatorClient: clients.OperatorClient,
+		infraLister:    clients.ConfigInformers.Config().V1().Infrastructures().Lister(),
+		versionGetter:  versionGetter,
+		targetVersion:  targetVersion,
+		eventRecorder:  eventRecorder,
+	}
+
+	// Populating all CSI driver operator ControllerManagers here simplifies
+	// the startup a lot
+	// - All necessary informers are populated.
+	// - CSO then just Run()s
+	//All ControllerManagers are nor running at this point! They will be started
+	c.populateCSIControllers(clients, resyncInterval, eventRecorder)
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(clients.OperatorClient).WithInformers(
+		clients.OperatorClient.Informer(),
+		clients.ConfigInformers.Config().V1().Infrastructures().Informer(),
+	).ToController("CSIDriverStarter", eventRecorder)
+}
+
+func (c *CSIDriverStarterController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	infrastructure, err := c.infraLister.Get(infraConfigName)
+	if err != nil {
+		return err
+	}
+
+	err = c.syncPlatform(ctx, infrastructure.Status.PlatformStatus.Type)
+	if err != nil {
+		return err
+	}
+	return nil
```

Comment:
nit: you can just return syncPlatform or, even better, move its contents here since it always return nil error

Reply:
moved syncPlatform to Sync()